File: frontend/src/api/transactions.ts
// src/api/transactions.ts
import { apiGet } from "./client";

export type NormalizedTxn = {
  id: string;
  source_event_id?: string | null;
  occurred_at?: string | null; // ISO
  date?: string | null;        // YYYY-MM-DD
  description?: string | null;
  amount: number;
  direction: "inflow" | "outflow";
  account?: string | null;
  category?: string | null;
  counterparty_hint?: string | null;
  merchant_key?: string | null;
  suggested_system_key?: string | null;
  suggested_category_id?: string | null;
  suggested_category_name?: string | null;
  suggestion_source?: string | null;
  confidence?: number | null;
  reason?: string | null;
};

export type TransactionsResponse = {
  business_id: string;
  name?: string;
  as_of: string;
  last_event_occurred_at?: string | null;
  count: number;
  transactions: NormalizedTxn[];
};

export type TransactionDetail = {
  business_id: string;
  source_event_id: string;
  raw_event: {
    source: string;
    source_event_id: string;
    payload: Record<string, any>;
    occurred_at: string;
    created_at: string;
    processed_at?: string | null;
  };
  normalized_txn: {
    source_event_id: string;
    occurred_at: string;
    date: string;
    description: string;
    amount: number;
    direction: "inflow" | "outflow";
    account: string;
    category_hint: string;
    counterparty_hint?: string | null;
    merchant_key?: string | null;
  };
  vendor_normalization: {
    canonical_name: string;
    source: string;
  };
  categorization?: {
    category_id: string;
    category_name: string;
    system_key?: string | null;
    account_id: string;
    account_name: string;
    source: string;
    confidence: number;
    note?: string | null;
    rule_id?: string | null;
    created_at: string;
  } | null;
  processing_assumptions: Array<{ field: string; detail: string }>;
  ledger_context?: {
    row: {
      source_event_id: string;
      occurred_at: string;
      date: string;
      description: string;
      vendor: string;
      amount: number;
      category: string;
      account: string;
      balance: number;
    };
    balance: number;
    running_total_in: number;
    running_total_out: number;
  } | null;
  audit_history: Array<{
    id: string;
    event_type: string;
    actor: string;
    reason?: string | null;
    before_state?: Record<string, any> | null;
    after_state?: Record<string, any> | null;
    rule_id?: string | null;
    created_at: string;
  }>;
  related_signals: Array<{
    signal_id: string;
    title?: string | null;
    severity?: string | null;
    status?: string | null;
    domain?: string | null;
    updated_at?: string | null;
    matched_on?: string | null;
    window?: Record<string, any> | null;
    facts?: Record<string, any> | null;
  }>;
};

export async function fetchTransactions(
  businessId: string,
  limit = 50,
  sourceEventIds?: string[],
  category?: string,
  direction?: "inflow" | "outflow"
): Promise<TransactionsResponse> {
  const params = new URLSearchParams();
  params.set("limit", String(limit));
  if (sourceEventIds?.length) params.set("source_event_ids", sourceEventIds.join(","));
  if (category) params.set("category", category);
  if (direction) params.set("direction", direction);

  return apiGet(`/demo/transactions/${businessId}?${params.toString()}`);
}

export async function fetchTransactionDetail(
  businessId: string,
  sourceEventId: string
): Promise<TransactionDetail> {
  return apiGet(`/api/transactions/${businessId}/${encodeURIComponent(sourceEventId)}`);
}

File: backend/app/api/routes/transactions.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.orm import Session

from backend.app.db import get_db
from backend.app.services.transaction_service import transaction_detail

router = APIRouter(prefix="/api/transactions", tags=["transactions"])


class RawEventOut(BaseModel):
    source: str
    source_event_id: str
    payload: Dict[str, Any]
    occurred_at: datetime
    created_at: datetime
    processed_at: Optional[datetime] = None


class NormalizedTxnOut(BaseModel):
    source_event_id: str
    occurred_at: datetime
    date: date
    description: str
    amount: float
    direction: str
    account: str
    category_hint: str
    counterparty_hint: Optional[str] = None
    merchant_key: Optional[str] = None


class VendorNormalizationOut(BaseModel):
    canonical_name: str
    source: str


class CategorizationOut(BaseModel):
    category_id: str
    category_name: str
    system_key: Optional[str] = None
    account_id: str
    account_name: str
    source: str
    confidence: float
    note: Optional[str] = None
    rule_id: Optional[str] = None
    created_at: datetime


class LedgerContextRowOut(BaseModel):
    source_event_id: str
    occurred_at: datetime
    date: date
    description: str
    vendor: str
    amount: float
    category: str
    account: str
    balance: float


class LedgerContextOut(BaseModel):
    row: LedgerContextRowOut
    balance: float
    running_total_in: float
    running_total_out: float


class AuditLogOut(BaseModel):
    id: str
    event_type: str
    actor: str
    reason: Optional[str] = None
    before_state: Optional[Dict[str, Any]] = None
    after_state: Optional[Dict[str, Any]] = None
    rule_id: Optional[str] = None
    created_at: datetime


class RelatedSignalOut(BaseModel):
    signal_id: str
    title: Optional[str] = None
    severity: Optional[str] = None
    status: Optional[str] = None
    domain: Optional[str] = None
    updated_at: Optional[datetime] = None
    matched_on: Optional[str] = None
    window: Optional[Dict[str, Any]] = None
    facts: Optional[Dict[str, Any]] = None
    recommended_actions: List[Dict[str, Any]] = []


class SuggestedCategoryOut(BaseModel):
    system_key: str
    category_id: str
    category_name: str
    source: str
    confidence: float
    reason: str


class RuleSuggestionOut(BaseModel):
    contains_text: str
    category_id: str
    category_name: str
    direction: Optional[str] = None
    account: Optional[str] = None


class TransactionDetailOut(BaseModel):
    business_id: str
    source_event_id: str
    raw_event: RawEventOut
    normalized_txn: NormalizedTxnOut
    vendor_normalization: VendorNormalizationOut
    categorization: Optional[CategorizationOut] = None
    suggested_category: Optional[SuggestedCategoryOut] = None
    rule_suggestion: Optional[RuleSuggestionOut] = None
    processing_assumptions: List[Dict[str, str]]
    ledger_context: Optional[LedgerContextOut] = None
    audit_history: List[AuditLogOut]
    related_signals: List[RelatedSignalOut]


@router.get("/{business_id}/{source_event_id}", response_model=TransactionDetailOut)
def get_transaction_detail(
    business_id: str,
    source_event_id: str,
    db: Session = Depends(get_db),
):
    """
    Audit-grade transaction detail by source_event_id.
    Includes raw ingestion, normalization assumptions, and audit log history tied to the event.
    """
    return transaction_detail(db, business_id, source_event_id)
